iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/digitaltwin/DigitalTwinClientTests.java

```diff
@@ -225,6 +225,36 @@ public void getDigitalTwinWithAzureSasCredential() {
         assertEquals(responseWithHeaders.body().getMetadata().getModelId(), E2ETestConstants.THERMOSTAT_MODEL_ID);
     }
 
+    @Test
+    public void azureSasCredentialTokenRenewal() throws Exception
+    {
+        IotHubConnectionString iotHubConnectionStringObj =
+            IotHubConnectionStringBuilder.createIotHubConnectionString(IOTHUB_CONNECTION_STRING);
+
+        // create a shared access signature that only lives for 5 seconds so that we can test renewing it
+        int tokenLifespanSeconds = 5;
+        int tokenLifespanMilliseconds = tokenLifespanSeconds * 1000;
+        IotHubServiceSasToken serviceSasToken = new IotHubServiceSasToken(iotHubConnectionStringObj, tokenLifespanSeconds);
+
+        AzureSasCredential sasCredential = new AzureSasCredential(serviceSasToken.toString());
+        digitalTwinClient = new DigitalTwinClient(iotHubConnectionStringObj.getHostName(), sasCredential);
+
+        // get a digital twin before the first shared access signature expires
+        // don't care about the return value, just checking that the request isn't unauthorized
+        digitalTwinClient.getDigitalTwin(deviceId, BasicDigitalTwin.class);
+
+        // wait so that the previous shared access signature expires
+        Thread.sleep(2 * tokenLifespanMilliseconds);
```

Comment:
how about a getTwin -> unauth -> getTwin test, asserting that the token actually did expire

Reply:
Good idea, I've reworked the tests a bit to deliberately cause a 401 error